File: leetcode/1046. Last Stone Weight/s3.py
# OJ: https://leetcode.com/problems/last-stone-weight/
# Author: github.com/punkfulw
# Time: O(NlogN)
# Space: O(1)
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        for i, s in enumerate(stones):
            stones[i] = -s
        heapify(stones)
        while len(stones) > 1:
            f, s = -heappop(stones), -heappop(stones)
            if f - s > 0:
                heappush(stones, s - f)
        if not stones:
            return 0
        return -stones[0]
            

File: leetcode/63. Unique Paths II/s3.cpp
// OJ: https://leetcode.com/problems/unique-paths-ii/
// Author: github.com/punkfulw
// Time: O(MN)
// Space: O(N)
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& grid) {
        int rec[100] {};
        int n = grid.size(), m = grid[0].size();
        rec[0] = 1;
        
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                if (grid[i][j])
                    rec[j] = 0;
                else if (j > 0)
                    rec[j] += rec[j - 1];
        return rec[m - 1];
    }
};
